workflow/util/plugins/plugin.go

```diff
@@ -0,0 +1,89 @@
+package plugin
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/argoproj/argo-workflows/v3/util/errors"
+)
+
+type Client struct {
+	Address string
+	client  http.Client
+	invalid map[string]bool
+	backoff wait.Backoff
+}
+
+func New(address string, timeout time.Duration, backoff wait.Backoff) Client {
+	return Client{
+		Address: address,
+		client: http.Client{
+			Timeout: timeout,
+		},
+		invalid: map[string]bool{},
+		backoff: backoff,
+	}
+}
+
+func (p *Client) Call(method string, args interface{}, reply interface{}) error {
+	if p.invalid[method] {
+		return nil
+	}
+	log := log.WithField("address", p.Address).WithField("method", method)
+	body, err := json.Marshal(args)
+	if err != nil {
+		return err
+	}
+	return retry.OnError(p.backoff, func(err error) bool {
+		log.WithError(err).Debug("Plugin returned error")
+		switch e := err.(type) {
+		case interface{ Temporary() bool }:
+			if e.Temporary() {
+				return true
+			}
+		}
+		return strings.Contains(err.Error(), "connection refused")
+	}, func() error {
+		log.Debug("Calling plugin")
+		req, err := http.NewRequest("POST", fmt.Sprintf("%s/api/v1/%s", p.Address, method), bytes.NewBuffer(body))
```

Comment:
Will the plugins be able to add other information to the request, e.g. user agent?

Reply:
no, not needed